### PR TITLE
Added get-status command

### LINKED ISSUE #78: Add a way for user to check state of deployment
User should be able to retrieve a status of his deployment (either detailed, or just a general status). 

This may be connected to defining a timeout on deployment or deployment parts (which would allow us to define a "timeout" status).

Use case
- user starts deployment as usual
- user checks deployment status via `kubectl exec k8s-appcontroller get-status`
- get-status binary prints out the deployment status

### PR DESCRIPTION
**Pull request by zefciu**
The 'get-status' command would print the information about the status of
deployment.
fixes #78

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/k8s-appcontroller/90)

### CODE REVIEW

**Comment by pigmej** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[Dockerfile, line 9 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT9-2dNAEXPN_jwKfqF:-KT9-2dNAEXPN_jwKfqG:b-w3ho61) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/Dockerfile#L9)):*
> ```Dockerfile
> 
> RUN apk --no-cache add git
> RUN go get ./... && go build -o kubeac cmd/kubeac/main.go && go build -o wrap cmd/wrap/main.go && go build -o bootstrap cmd/bootstrap/main.go && go build -o get-status cmd/get-status/main.go && mv kubeac /usr/bin/kubeac && mv wrap /usr/bin/wrap && mv bootstrap /usr/bin/bootstrap && mv get-status /usr/bin/get-status && mkdir -p /opt/kubeac && mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests && rm -fr /go
> ```

I'd prefer it to have it integrated with kubeac binary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

File: scheduler/scheduler.go [hunk 1]
```diff
@@ -250,6 +276,7 @@ func BuildDependencyGraph(c client.Interface, sel labels.Selector) (DependencyGr
 			depGraph[child].Requires, depGraph[parent])
 
 		depGraph[child].Meta[parent] = d.Meta
+		depGraph[parent].Meta[parent] = d.Meta
```
Comment (jellonek) [inline scheduler/scheduler.go]: Hmmm.... shouldn't there be `.Meta[child]`?
Looks like this is overwriting something.

File: Dockerfile [hunk 2]
```diff
@@ -6,7 +6,7 @@ COPY . /go/src/github.com/Mirantis/k8s-AppController
 WORKDIR /go/src/github.com/Mirantis/k8s-AppController
 
 RUN apk --no-cache add git
-RUN go get ./... && go build -o kubeac cmd/kubeac/main.go && go build -o wrap cmd/wrap/main.go && go build -o bootstrap cmd/bootstrap/main.go && mv kubeac /usr/bin/kubeac && mv wrap /usr/bin/wrap && mv bootstrap /usr/bin/bootstrap && mkdir -p /opt/kubeac && mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests && rm -fr /go
+RUN go get ./... && go build -o kubeac cmd/kubeac/main.go && go build -o wrap cmd/wrap/main.go && go build -o bootstrap cmd/bootstrap/main.go && go build -o get-status cmd/get-status/main.go && mv kubeac /usr/bin/kubeac && mv wrap /usr/bin/wrap && mv bootstrap /usr/bin/bootstrap && mv get-status /usr/bin/get-status && mkdir -p /opt/kubeac && mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests && rm -fr /go
```
Comment (nebril) [inline Dockerfile]: Please add the command you are adding as a subcommand of kubeac binary to keep image size low. You can use `cmd` directory as a reference.

File: scheduler/scheduler.go (hunk 1 above)
Comment (nebril) [inline scheduler/scheduler.go]: +1, please explain what are you trying to do here.

Comment (zefciu) [inline scheduler/scheduler.go]: We discussed it already. If you look at the way get-status is iterating over the tree, you will see that I need the metadata to be accessible on parent to be able to check the parent's status. Do you want me to put comments there?

Comment (jellonek) [inline scheduler/scheduler.go]: But with multiple children same You will in result have only last value for this parent. Am I right?
IMO Ilya should also take a look on this.

Btw. I hate inline comments (they are quite good markers that code needs to be refactored into simpler one, but better understandable) - but this is clearly a good place for such comment.

Comment (zefciu) [inline scheduler/scheduler.go]: OK. After discussion I admit, that this is wrong and a result of my misunderstanding. I will correct it.

**Comment by zefciu** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[scheduler/scheduler.go, line 138 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KTiBspBp_wSFkHB30Zd:-KTiBspCUIa1SXau4Jf8:b-t06o5l) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/scheduler/scheduler.go#L138)):*
> Found resource

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

**Comment by zefciu** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[scheduler/scheduler.go, line 138 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KTiBspBp_wSFkHB30Zd:-KTilV3L2YepqaoJdTl-:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/scheduler/scheduler.go#L138)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> > Found resource
</details>
Done.

---

*[scheduler/scheduler.go, line 279 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT8zOOHyr3c_w0pPkcr-r1-279:-KTilNr39L43xgCGjxso:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/scheduler/scheduler.go#L279)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> OK. After discussion I admit, that this is wrong and a result of my misunderstanding. I will correct it.
</details>
Done.

---

*[Dockerfile, line 9 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT8zOOHyr3c_w0pPkcq-r1-9:-KTilP7UK5yf6mt5R2sa:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/Dockerfile#L9)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> Please add the command you are adding as a subcommand of kubeac binary to keep image size low. You can use `cmd` directory as a reference.
</details>
Done.

---

*[Dockerfile, line 9 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT9-2dNAEXPN_jwKfqF:-KTilPWjhV2QcbfViObj:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/Dockerfile#L9)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> I'd prefer it to have it integrated with kubeac binary.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

**Comment by zefciu** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[scheduler/scheduler.go, line 138 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KTiBspBp_wSFkHB30Zd:-KTilfkvSSEFK0qDE7Pq:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/scheduler/scheduler.go#L138)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> Done.
</details>
Done.

---

*[scheduler/scheduler.go, line 279 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT8zOOHyr3c_w0pPkcr-r1-279:-KTilgmvinAfQJ2fYXsT:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/scheduler/scheduler.go#L279)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> Done.
</details>
Done.

---

*[Dockerfile, line 9 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT8zOOHyr3c_w0pPkcq-r1-9:-KTiliVYkzT9yogTLmK9:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/Dockerfile#L9)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> Done.
</details>
Done.

---

*[Dockerfile, line 9 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT9-2dNAEXPN_jwKfqF:-KTiljEQ7pk8bixIASB8:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/fbb705be37d0441fe0f4089ad67d3497f3493ec8/Dockerfile#L9)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

**Comment by nebril** [comment]
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[cmd/get-status.go, line 19 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KTimxvA-uKdsxC59VNO:-KTimxvA-uKdsxC59VNP:byx47l4) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/c3eccec77d50847c4f3dcf8d8ce96f98be9097e5/cmd/get-status.go#L19)):*
> ```Go
> 	var err error
> 
> 	concurrency, err := cmd.Flags().GetInt("concurrency")
> ```

Please remove this flag, it's not needed int this command.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

**Comment by zefciu** [comment]
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[cmd/get-status.go, line 19 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KTimxvA-uKdsxC59VNO:-KTitbkv27OYI895evck:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/c3eccec77d50847c4f3dcf8d8ce96f98be9097e5/cmd/get-status.go#L19)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> Please remove this flag, it's not needed int this command.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

File: cmd/get-status.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	"os"
+
+	"github.com/Mirantis/k8s-AppController/client"
+	"github.com/Mirantis/k8s-AppController/scheduler"
+	"github.com/spf13/cobra"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+// GetStatus is a command that prints the deployment status
+func GetStatus(cmd *cobra.Command, args []string) {
```
Comment (nebril) [inline cmd/get-status.go]: This function doesn't need to be exported.

File: scheduler/scheduler.go [hunk 4]
```diff
@@ -459,3 +497,44 @@ func assignVertex(vertex, root *ScheduledResource, assigned map[string]bool, com
 		}
 	}
 }
+
+// GetStatus generates data for getting the status of deployment. Returns
+// a DeploymentStatus and a human readable report string
+// TODO: Allow for other formats of report (e.g. json for visualisations)
+func (graph DependencyGraph) GetStatus() (DeploymentStatus, string, error) {
```
Comment (nebril) [inline scheduler/scheduler.go]: This function needs to be tested.

**Comment by zefciu** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[scheduler/scheduler.go, line 504 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KT8zOOHyr3c_w0pPkcr-r3-504:-KTnZJip5r0M33X2SWRH:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9adee95dd7823fd5fe4b71f56fd4288c968ea26f/scheduler/scheduler.go#L504)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> This function needs to be tested.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

**Comment by zefciu** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cmd/get-status.go, line 15 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90#-KTijmAM19V3r_K624H0-r3-15:-KTnZgkYyJl3_mgkwyvJ:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/9adee95dd7823fd5fe4b71f56fd4288c968ea26f/cmd/get-status.go#L15)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> This function doesn't need to be exported.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/90)*

**Pull request merged**